deduction_rules.py

```diff
@@ -107,21 +106,129 @@ def search_hidden_pairs(elems):
         double_pairs = set([pair for pair in two_potential_place.values() if list(two_potential_place.values()).count(pair) == 2])
         return {pair: tuple([k for k,v in two_potential_place.items() if v==pair]) for pair in double_pairs}
 
-    def search_and_ban_in_subset(cells_to_check):
+    def search_and_ban_in_subset(cells_to_check, section):
         made_deduction = False
         allowed_numbers = _allowed_numbers(sudoku,cells_to_check)
         pairs = search_hidden_pairs(allowed_numbers)
         for pair,except_nums in pairs.items():
-            current_hidden_cell0 = cells_to_check[pair[0]]
-            current_hidden_cell1 = cells_to_check[pair[1]]
-            cells_used = sudoku.allowed[current_hidden_cell0[0]][current_hidden_cell0[1]].notNones()+sudoku.allowed[current_hidden_cell1[0]][current_hidden_cell1[1]].notNones()
-            made_deduction |= _ban_numbers(sudoku, current_hidden_cell0, filter(lambda x: x not in except_nums, allowed_numbers[pair[0]]), "hidden_pair",cells_used)
-            made_deduction |= _ban_numbers(sudoku, current_hidden_cell1, filter(lambda x: x not in except_nums, allowed_numbers[pair[1]]), "hidden_pair",cells_used)
+            cell0 = cells_to_check[pair[0]]
+            cell1 = cells_to_check[pair[1]]
+            used_nums = set(range(1,10))-set(except_nums)
+            cells_used = []
+            for cell in cells_to_check:
+                if cell not in (cell0, cell1):
+                    cells_used += [sudoku.allowed[cell[0]][cell[1]][n] for n in used_nums if n in _allowed_numbers(sudoku,[cell])[0]
```

Comment:
Missing `]` at the end of the line. Also see same indexing bug in `hidden_triple`.